### DISCUSSION TITLE
Splink is not working with a 1.5M rows dataset

### DISCUSSION THREAD

**Discussion by robertoelarte** [moj-analytical-services/splink#674] [Q&A]
Hi everyone!

I would like to tell you about the problem we have with Splink. We are using a 1.5M rows dataset with some variables about customers. We have, for example, the complete name, complete address, phone number, zip code, etc. Then, we are trying to use Splink in order to calculate the clusters because we know we have duplicated customers in the database.

The issue is that the algorithm never ends. We are using the SparkLinker so that it can work with a huge dataset (as the authors explain in the repo), but it does not work at all. It seems that it is stuck in the driver, even if we activate the "pyarrow". We have seen that although it is called "SparkLinker", inside the function you can find a "toPandas". Then, we have seen that the algorithm gets stuck in that line of code because it cannot process all the previous joins with the "toPandas" part.

We are using blocking rules in order to let the machine to compute it well. Here we show the settings we are using:

```
settings_test = {
        "link_type": "dedupe_only",
        "blocking_rules_to_generate_predictions": [
            {"blocking_rule": "l.complete_name = r.complete_name", "salting_partitions": 3},
        ],
        "comparisons": [
            cl.levenshtein_at_thresholds("complete_address"),
            cl.exact_match("phone_number"),
        ],
    }

linker = SparkLinker(dataset, settings_test)

linker.estimate_u_using_random_sampling(target_rows=1e6)
blocking_rule_for_training = "l.complete_name = r.complete_name"
linker.estimate_parameters_using_expectation_maximisation(blocking_rule_for_training)
df_predict = linker.predict()
```

We have a machine with the following specifications:

- 6 workers, 56GB memory, 16 cores each one
- driver: 56gb memory, 16 cores

So we do not have a small machine so to speak.

Do you know what are we doing wrong? The authors of the package say that the algorithm should work with a 100M rows dataset.

Thank you very much for your help!

**Comment by MirandaLochner** [comment]
Hello! I am also trying to run a 1.5million row dataset with Splink 3.0. We are able to get the df_predictions to write to our data lake but the clusters/connected component piece has not worked for us on 1.5 million rows. I'll let you know if we figure out why. 

For your example, I would estimate if you block on complete_name how many possible combinations that results in. Something like:
SELECT COUNT(*)
FROM table t1
INNER JOIN table t2
ON t1.complete_name = t2.complete_name 
WHERE t1.uniqueIdColumn <  t2.uniqueIdColumn

Once we got our count down to around 9 million for a 1.5 million row dataset it ran through the prediction step fast.

**Reply by RobinL** [reply]
You can also use this to compute the number of comparisons generated by a blocking rule.  https://moj-analytical-services.github.io/splink/linker.html#splink.linker.Linker.count_num_comparisons_from_blocking_rule

This is really the fundamental performance constraint - you'll want to keep the number of rows generated reasonably low, say less than 100 million. If that works you can always scale up further by relaxing the blocking rule

**Comment by RobinL** [comment]
I've just written a couple more topic guides that hopefully help:

https://moj-analytical-services.github.io/splink/topic_guides/drivers_of_performance.html
https://moj-analytical-services.github.io/splink/topic_guides/optimising_spark.html
